jdbc-core/src/main/java/com/salesforce/datacloud/jdbc/core/DataCloudStatement.java

```diff
@@ -209,33 +211,42 @@ public void setMaxRows(int max) {}
     @Override
     public void setEscapeProcessing(boolean enable) {}
 
-    protected Duration resolveQueryTimeout(Duration timeout) {
-        if (timeout == null) {
-            return getQueryTimeoutDuration();
-        }
-
-        if (timeout.isZero() || timeout.isNegative()) {
-            return Duration.ofSeconds(DEFAULT_QUERY_TIMEOUT);
+    /**
+     * This translates the user visible query timeout behavior (zero is infinite) to the internal expectation of the code
+     * (which would interpret zero as literal zero timeout). Thus zero is translated to a practically infinite timeout by using
+     * a large second count.
+     *
+     * Note: This behavior should be fixed in the future by changing the internal code to interpret zero as not setting any timeout at all.
+     * @return The effective query timeout duration.
+     */
+    protected Duration getEffectiveQueryTimeoutDuration() {
+        if (statementProperties.getQueryTimeout() == Duration.ZERO) {
+            return Duration.ofSeconds(Constants.INFINITE_QUERY_TIMEOUT);
         }
-
-        return timeout;
-    }
-
-    protected Duration getQueryTimeoutDuration() {
-        return Duration.ofSeconds(getQueryTimeout());
+        return statementProperties.getQueryTimeout();
     }
 
+    /**
+     * @return The query timeout for this statement in seconds.
+     */
     @Override
     public int getQueryTimeout() {
-        return queryTimeout;
+        return (int) statementProperties.getQueryTimeout().getSeconds();
     }
 
+    /**
+     * Sets the query timeout for this statement. A zero or negative value is interpreted as infinite timeout.
+     * @param seconds The query timeout in seconds.
+     */
     @Override
     public void setQueryTimeout(int seconds) {
```

Comment:
Rather use proper types

```suggestion
    public void setQueryTimeout(Duration seconds) {
```